From e935b2a64dcfad8fddfac86d08659a62436778e9 Mon Sep 17 00:00:00 2001
From: Evoke Zhang <evoke.zhang@amlogic.com>
Date: Thu, 28 Aug 2014 10:08:35 +0800
Subject: [PATCH 4798/5965] PD#95667: lcd driver: fix system halt when access
 edp registers if the edp driver didn't init in uboot

Squashed commit of the following:

commit 44cd0d19d111c35814189453b596c96f755b793d
Author: Evoke Zhang <evoke.zhang@amlogic.com>
Date:   Fri Aug 8 16:33:58 2014 +0800

    PD#95667: lcd_driver: remove bl_status flag

    Change-Id: I7502214fdd4279940340a29338137838b435f4fd

commit 1074f1bb151b0dfd566bf5ea605f3a5ba7234bad
Author: Evoke Zhang <evoke.zhang@amlogic.com>
Date:   Fri Aug 8 14:10:30 2014 +0800

    PD#95667: lcd driver: fix system halt when access edp registers if the edp driver didn't init in uboot

    Change-Id: Ib22dcc504d4852c9e691792cec7a778667409a25

Change-Id: I2e0e6d14caf8eee24de238c1fbcac0ebae971152
---
 arch/arm/mach-meson8/lcd/lcd_config.c  | 15 ++++++++++-----
 arch/arm/mach-meson8/lcd/lcd_config.h  |  2 +-
 drivers/amlogic/display/vout/lcdoutc.c |  6 +-----
 3 files changed, 12 insertions(+), 11 deletions(-)

diff --git a/arch/arm/mach-meson8/lcd/lcd_config.c b/arch/arm/mach-meson8/lcd/lcd_config.c
index f40576211242..bcf7a839fd2f 100755
--- a/arch/arm/mach-meson8/lcd/lcd_config.c
+++ b/arch/arm/mach-meson8/lcd/lcd_config.c
@@ -559,11 +559,6 @@ static void vclk_set_lcd(int lcd_type, unsigned long pll_reg, unsigned long vid_
 			WRITE_LCD_CBUS_REG_BITS(HHI_EDP_TX_PHY_CNTL0, edp_div0_sel, 20, 4);	//set edptx_phy_clk_div0
 			WRITE_LCD_CBUS_REG_BITS(HHI_EDP_TX_PHY_CNTL0, edp_div1_sel, 24, 3);	//set edptx_phy_clk_div1
 			WRITE_LCD_CBUS_REG_BITS(HHI_EDP_TX_PHY_CNTL0, 1, 5, 1);	//enable divider N, for vid_pll2_in
-			
-			if (IS_MESON_M8_CPU)
-				WRITE_LCD_CBUS_REG(HHI_EDP_APB_CLK_CNTL, (1 << 7) | (2 << 0));	//fclk_div5---fixed 510M, div to 170M, edp apb clk
-			else if (IS_MESON_M8M2_CPU)
-				WRITE_LCD_CBUS_REG(HHI_EDP_APB_CLK_CNTL_M8M2, (1 << 7) | (2 << 0));	//fclk_div5---fixed 510M, div to 170M, edp apb clk
 			break;
 		case LCD_DIGITAL_LVDS:
 		case LCD_DIGITAL_TTL:
@@ -602,6 +597,8 @@ static void vclk_set_lcd(int lcd_type, unsigned long pll_reg, unsigned long vid_
 	WRITE_LCD_CBUS_REG_BITS(HHI_VIID_CLK_CNTL, 0, 15, 1);  //release soft reset
 	udelay(5);
 	
+	if (IS_MESON_M8M2_CPU)
+		WRITE_LCD_CBUS_REG_BITS(HHI_VID_CLK_CNTL2, 1, 3, 1);	//enable CTS_ENCL clk gate, new added in m8m2
 	spin_unlock_irqrestore(&lcd_clk_lock, flags);
 }
 
@@ -1118,6 +1115,8 @@ static void _disable_lcd_driver(Lcd_Config_t *pConf)
 
     WRITE_LCD_REG(ENCL_VIDEO_EN, 0);	//disable encl
 
+    if (IS_MESON_M8M2_CPU)
+        WRITE_LCD_CBUS_REG_BITS(HHI_VID_CLK_CNTL2, 0, 3, 1);	//disable CTS_ENCL clk gate, new added in m8m2
     WRITE_LCD_CBUS_REG_BITS(HHI_VIID_CLK_CNTL, 0, 0, 5);	//close vclk2 gate: 0x104b[4:0]
 
     WRITE_LCD_CBUS_REG_BITS(HHI_VIID_DIVIDER_CNTL, 0, 16, 1);	//close vid2_pll gate: 0x104c[16]
@@ -1751,6 +1750,12 @@ static void lcd_control_config_pre(Lcd_Config_t *pConf)
             set_mipi_dsi_control_config(pConf);
             break;
         case LCD_DIGITAL_EDP:
+            //prepare EDP_APB_CLK to access EDP registers
+            if (IS_MESON_M8_CPU)
+                WRITE_LCD_CBUS_REG(HHI_EDP_APB_CLK_CNTL, (1 << 7) | (2 << 0));      //fclk_div5---fixed 510M, div to 170M, edp apb clk
+            else if (IS_MESON_M8M2_CPU)
+                WRITE_LCD_CBUS_REG(HHI_EDP_APB_CLK_CNTL_M8M2, (1 << 7) | (2 << 0)); //fclk_div5---fixed 510M, div to 170M, edp apb clk
+
             ss_level = ((ss_level > 0) ? 1 : 0);
             select_edp_link_config(pConf);
             if (pConf->lcd_control.edp_config->link_adaptive == 1) {
diff --git a/arch/arm/mach-meson8/lcd/lcd_config.h b/arch/arm/mach-meson8/lcd/lcd_config.h
index 83b4a0dfe597..af3cb8c53388 100755
--- a/arch/arm/mach-meson8/lcd/lcd_config.h
+++ b/arch/arm/mach-meson8/lcd/lcd_config.h
@@ -7,7 +7,7 @@
 //lcd driver version
 //**********************************
 #define LCD_DRV_TYPE      "c8"
-#define LCD_DRV_DATE      "20140805"
+#define LCD_DRV_DATE      "20140808"
 //**********************************
 
 /* for GAMMA_CNTL_PORT */
diff --git a/drivers/amlogic/display/vout/lcdoutc.c b/drivers/amlogic/display/vout/lcdoutc.c
index 7d20f0d3b471..7d47382913f2 100755
--- a/drivers/amlogic/display/vout/lcdoutc.c
+++ b/drivers/amlogic/display/vout/lcdoutc.c
@@ -95,7 +95,6 @@ typedef struct {
 static lcd_dev_t *pDev = NULL;
 static struct class *gamma_debug_class = NULL;
 static Bool_t data_status = ON;
-static int bl_status = ON;
 
 static inline void lcd_mdelay(int n)
 {
@@ -145,17 +144,14 @@ static void lcd_setup_gamma_table(Lcd_Config_t *pConf, unsigned int rgb_flag)
 static void backlight_power_ctrl(Bool_t status)
 {
 	if( status == ON ){
-		if ((data_status == OFF) || (bl_status == ON))
+		if (data_status == OFF)
 			return;
 		bl_power_on(LCD_BL_FLAG);
 	}
 	else{
-		if (bl_status == OFF)
-			return;
 		bl_power_off(LCD_BL_FLAG);
 	}
 	lcd_print("%s(%s): data_status=%s\n", __FUNCTION__, (status ? "ON" : "OFF"), (data_status ? "ON" : "OFF"));
-	bl_status = status;
 }
 
 static int lcd_power_ctrl(Bool_t status)
-- 
2.19.0

